squad/core/tasks/notification.py

```diff
@@ -9,13 +9,30 @@
 
 @celery.task
 def maybe_notify_project_status(status_id):
+    """
+        This is invoked for every new TestRun a build receives.
+        Notifications tasks work like the following:
+
+        1. First notification attempt with `maybe_notify_project_status`,
+           it uses the attribute `Project.wait_before_notification` which
+           waits for this number of seconds AFTER `Build.datetime`. This means
+           that if the build date keeps changing, notification will delay accordingly.
+           NOTE: the notification will be sent only if `ProjectStatus.finished=True`,
+           even after timeout expiration.
+
+        2. Second notifcation attempt is achieved at `notification_timeout`, which
+           waits for an absolute number of seconds, e.g. `Project.notification_timeout`,
+           before trying to send the notification. This serves as a fallback option to
+           send out notifications, even if `ProjectStatus.finished=False`.
+    """
     projectstatus = ProjectStatus.objects.get(pk=status_id)
     build = projectstatus.build
     project = build.project
 
-    timeout = project.notification_timeout
-    if timeout:
-        notification_timeout.apply_async(args=[status_id], countdown=timeout)
+    if projectstatus.notified_on_timeout is None and project.notification_timeout:
+        notification_timeout.apply_async(args=[status_id], countdown=project.notification_timeout)
+        projectstatus.notified_on_timeout = False
+        projectstatus.save()
```

Comment:
there is a race condition between fetching `projectstatus` from the DB and getting `notified_on_timeout is None`, and saving `notified_on_timeout = False`. you probably want to wrap this  in an atomic transation.

Reply:
You're right. I think you suggested an atomic transaction on something else also related to tasks. I'll try to come up with it so that this issue can be addressed as well: https://github.com/Linaro/squad/issues/674